### PR TITLE
Add tests for vm Status transition metrics windows

### PR DESCRIPTION
**Pull request by OhadRevah**
##### Short description:
Add tests for vm status transition metircs for windows vms. The following metrics added:
kubevirt_vm_running_status_last_transition_timestamp_seconds kubevirt_vm_error_status_last_transition_timestamp_seconds kubevirt_vm_migrating_status_last_transition_timestamp_seconds kubevirt_vm_non_running_status_last_transition_timestamp_seconds kubevirt_vm_starting_status_last_transition_timestamp_seconds
##### jira-ticket:
https://issues.redhat.com/browse/CNV-63356

### CODE REVIEW

File: tests/observability/metrics/utils.py [hunk 1]
```diff
@@ -1477,3 +1478,51 @@ def get_vmi_guest_os_kernel_release_info_metric_from_vm(
...
+    LOGGER.info(f"Check VMI metric expected: {expected_vmi_count}")
+    samples = TimeoutSampler(
+        wait_timeout=100,
+        sleep=5,
+        func=check_vmi_metric,
+        prometheus=prometheus,
+    )
+    for sample in samples:
+        if sample == expected_vmi_count:
+            return True
```
[the remark this answers is not in the document]
Comment (OhadRevah) [reply tests/observability/metrics/utils.py:1528]: Good point just need to add try and expect for the for loop, the excpetion should be TimeoutExpiredError.

File: tests/observability/metrics/test_vms_metrics.py [hunk 2]
```diff
@@ -228,27 +83,72 @@ def test_vm_migrating_status_metrics(
...
+
+    @pytest.mark.polarion("CNV-11980")
+    def test_vm_migrating_status_metrics_windows(
+        self, skip_if_no_common_cpu, prometheus, windows_vm_for_test, vm_metric_1_vmim
+    ):
+        check_vm_last_transition_metric_value(
+            prometheus=prometheus,
+            metric=KUBEVIRT_VM_MIGRATING_STATUS_LAST_TRANSITION_TIMESTAMP_SECONDS,
+            vm=windows_vm_for_test,
+        )
```
[the remark this answers is not in the document]
Comment (OhadRevah) [reply tests/observability/metrics/test_vms_metrics.py:133]: Good one

File: tests/observability/metrics/conftest.py [hunk 3]
```diff
@@ -1200,3 +1214,108 @@ def migration_succeeded_scope_class(vm_migration_metrics_vmim_scope_class):
...
+@pytest.fixture()
+def windows_vm_vmim(windows_vm_for_test):
+    with VirtualMachineInstanceMigration(
+        name="vm-metric-1-vmim",
+        namespace=windows_vm_for_test.namespace,
+        vmi_name=windows_vm_for_test.vmi.name,
+    ) as vmim:
+        vmim.wait_for_status(status=vmim.Status.RUNNING, timeout=TIMEOUT_3MIN)
+        yield
+
```
[the remark this answers is not in the document]
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py:1304]: Good one.

**Comment by openshift-virtualization-qe-bot-2** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

File: tests/observability/metrics/conftest.py [hunk 4]
```diff
@@ -1144,3 +1186,109 @@ def migration_succeeded_scope_class(vm_migration_metrics_vmim_scope_class):
     vm_migration_metrics_vmim_scope_class.wait_for_status(
         status=vm_migration_metrics_vmim_scope_class.Status.SUCCEEDED, timeout=TIMEOUT_5MIN
     )
+
+
+@pytest.fixture()
+def stopped_vm_metric_1(vm_metric_1):
+    vm_metric_1.stop(wait=True)
+
```
[the remark this answers is not in the document]
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py:1180]: I don't need the vm to run again.

File: tests/observability/metrics/test_vms_metrics.py [hunk 5]
```diff
@@ -231,27 +86,70 @@ def test_vm_migrating_status_metrics(
...
+    @pytest.mark.polarion("CNV-11980")
+    def test_vm_migrating_status_metrics_windows(
+        self,
+        skip_if_no_common_cpu,
+        prometheus,
+        windows_vm_for_test,
+        windows_vm_with_low_bandwidth_migration_policy,
+        migration_policy_with_bandwidth,
+        windows_vm_vmim,
+    ):
```
[the remark this answers is not in the document]
Comment (OhadRevah) [reply tests/observability/metrics/test_vms_metrics.py:137]: I need to add the low bandwidth migration policy to the vm because windows vm takes a lot of time to deploy, so I am using the same fixture that creates the windows vm and just adding the needed label instead creating new one and wait a lot of time. so these fixtures are needed.

File: tests/observability/metrics/conftest.py [hunk 6]
```diff
@@ -1144,3 +1186,109 @@ def migration_succeeded_scope_class(vm_migration_metrics_vmim_scope_class):
...
+
+@pytest.fixture()
+def stopped_windows_vm(windows_vm_for_test):
+    windows_vm_for_test.stop(wait=True)
+
```
[the remark this answers is not in the document]
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py:1185]: I don't need the vm to run after this fixture.

File: tests/observability/metrics/conftest.py [hunk 7]
```diff
@@ -1055,6 +1057,47 @@ def windows_vm_for_test(namespace, unprivileged_client):
...
+            windows_vm_for_test: {
+                "spec": {
+                    "template": {"spec": {"nodeSelector": get_node_selector_dict(node_selector="non-existent-node")}}
+                }
+            }
+        }
+    ):
+        windows_vm_for_test.restart()
+        windows_vm_for_test.wait_for_specific_status(status=VirtualMachine.Status.ERROR_UNSCHEDULABLE)
+        yield
```
[the remark this answers is not in the document]
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py:1077]: It will better to change the scope of this fixture to function so it will be reverted in the teardown of the test using it.

**Comment by openshift-virtualization-qe-bot-2** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

**Comment by openshift-virtualization-qe-bot-3** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

**Comment by openshift-virtualization-qe-bot-3** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

**Comment by openshift-virtualization-qe-bot-3** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

File: tests/observability/metrics/conftest.py [hunk 8]
```diff
@@ -1144,3 +1172,99 @@ def migration_succeeded_scope_class(vm_migration_metrics_vmim_scope_class):
...
+        additional_labels=MIGRATION_POLICY_VM_LABEL,
+        cpu_model=cluster_common_node_cpu,
+    ) as vm:
+        running_vm(vm=vm, wait_for_interfaces=False, check_ssh_connectivity=False)
+        yield vm
+
+
+@pytest.fixture()
+def vm_metric_1_vmim(vm_metric_1):
+    with VirtualMachineInstanceMigration(
```
Comment (rnetser) [inline tests/observability/metrics/conftest.py:1234]: please add `client` in all ocp resources

**Review by rnetser — CHANGES_REQUESTED** [review]

**Comment by openshift-virtualization-qe-bot-2** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

File: tests/observability/metrics/conftest.py [hunk 9]
```diff
@@ -1144,3 +1174,100 @@ def migration_succeeded_scope_class(vm_migration_metrics_vmim_scope_class):
...
+@pytest.fixture()
+def windows_vm_vmim(migration_policy_with_bandwidth, windows_vm_for_test):
+    with VirtualMachineInstanceMigration(
+        name="windows-vm-metric-1-vmim",
+        namespace=windows_vm_for_test.namespace,
+        vmi_name=windows_vm_for_test.vmi.name,
+    ) as vmim:
+        vmim.wait_for_status(status=vmim.Status.RUNNING, timeout=TIMEOUT_20MIN)
+        windows_vm_for_test.wait_for_specific_status(status=windows_vm_for_test.Status.MIGRATING)
+        yield
```
[the remark this answers is not in the document]
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py]: Done

**Comment by openshift-virtualization-qe-bot-3** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

File: tests/observability/metrics/conftest.py [hunk 10]
```diff
@@ -1055,9 +1057,37 @@ def windows_vm_for_test(namespace, unprivileged_client):
...
+                    "template": {"spec": {"nodeSelector": get_node_selector_dict(node_selector="non-existent-node")}}
+                }
+            }
+        }
+    ):
+        windows_vm_for_test.restart()
+        windows_vm_for_test.wait_for_specific_status(status=VirtualMachine.Status.ERROR_UNSCHEDULABLE)
+        yield windows_vm_for_test
+        windows_vm_for_test.restart()
+        windows_vm_for_test.wait_for_specific_status(status=VirtualMachine.Status.RUNNING)
```
Comment (dshchedr) [inline tests/observability/metrics/conftest.py:1079]: I guess these two lines should be moved 1 step left, outside of `with ResourceEditor()`

but another point - not clear why you want to restart the VM after reverting changes back? you anyway restart it in `windows_vm_with_low_bandwidth_migration_policy` fixture

**Review by dshchedr — CHANGES_REQUESTED** [review]

File: tests/observability/metrics/conftest.py (hunk 10 above)
Comment (dshchedr) [reply tests/observability/metrics/conftest.py]: that's because in next fixture you have ` windows_vm_for_test.restart(wait=True)`
when you set "wait=True" - it looks for VMI, but you don't have vmi because in previous test VM was in failed state

with you current approach you are restarting VM two times: during teardown of "error" test and setup of "migration", it does not make sense.

I guess you can move these two lines to the fixture `windows_vm_with_low_bandwidth_migration_policy` - put them instead of `windows_vm_for_test.restart(wait=True)`

or even better use this lines (but anyway need to move them):
```
windows_vm_for_test.restart()
running_vm(vm=windows_vm_for_test)
```

Comment (dshchedr) [reply tests/observability/metrics/conftest.py]: I don't see changes
you need to remove extra restart from fixture `windows_vm_for_test_in_error_state`, these lines:

```
    windows_vm_for_test.restart()
    windows_vm_for_test.wait_for_specific_status(status=VirtualMachine.Status.RUNNING)
```

File: tests/observability/metrics/conftest.py [hunk 11]
```diff
@@ -1056,9 +1058,37 @@ def windows_vm_for_test(namespace, unprivileged_client):
...
+                    "template": {"spec": {"nodeSelector": get_node_selector_dict(node_selector="non-existent-node")}}
+                }
+            }
+        }
+    ):
+        windows_vm_for_test.restart()
+        windows_vm_for_test.wait_for_specific_status(status=VirtualMachine.Status.ERROR_UNSCHEDULABLE)
+        yield windows_vm_for_test
+    windows_vm_for_test.restart()
+    windows_vm_for_test.wait_for_specific_status(status=VirtualMachine.Status.RUNNING)
```
Comment (dshchedr) [inline tests/observability/metrics/conftest.py:1080]: these line not needed

**Review by dshchedr — CHANGES_REQUESTED** [review]

File: tests/observability/metrics/conftest.py [hunk 12]
```diff
@@ -1058,9 +1059,38 @@ def windows_vm_for_test(namespace, unprivileged_client):
...
+                    "template": {"spec": {"nodeSelector": get_node_selector_dict(node_selector="non-existent-node")}}
+                }
+            }
+        }
+    ):
+        windows_vm_for_test.restart()
+        windows_vm_for_test.wait_for_specific_status(status=VirtualMachine.Status.ERROR_UNSCHEDULABLE)
+        yield windows_vm_for_test
+    windows_vm_for_test.restart()
+    windows_vm_for_test.wait_for_specific_status(status=windows_vm_for_test.Status.RUNNING, timeout=TIMEOUT_10MIN)
```
Comment (dshchedr) [inline tests/observability/metrics/conftest.py:1081]: you still have extra restart here (and even more - now you added extra 10 minutes for it)
I don't understand why you need it because you anyway restart the VM in another fixture

**Review by dshchedr — CHANGES_REQUESTED** [review]

File: tests/observability/metrics/conftest.py (hunk 12 above)
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py:1081]: Without this two lines the tests failing, it has to be done, when I removed this lines the tests after this one with windows vm failed because fixtures could not find this vm anymore and failed at setup, I have to do restart in the teardown of this fixture, and then make sure it is in running state, it takes more time then usual linux vm to get this status, less then 10 mins made the tests flaky, I tried a lot of times and only with this lines the tests pass.

Comment (dshchedr) [reply tests/observability/metrics/conftest.py:1081]: ok, restart once at the end anyway better than multiple times:

create VM with the label for migration policy, but don't create migration policy itself yet

with this order you will have only one restart at the end:

1. test_vm_running_status_metrics_windows (create + run vm)
2. test_vm_migrating_status_metrics_windows (no need to restart VM: create migration policy, migrate VM, delete migration policy)
3. test_vm_non_running_status_metrics_windows (stop VM)
4. test_vm_error_status_metrics_windows (edit vm, start - it should be quick because we don't wait when it is active, restart and wait when running during teardown)

Comment (OhadRevah) [reply tests/observability/metrics/conftest.py:1081]: I need the label for migration policy only for one test,  so adding the label with the creation of the windows will have to be in new fixture so I will not need to edit it with `Resource Editor` and restart again.
Creating a windows vm takes a lot of time, this is why I have created the windows vm with module scope so all the tests that needs windows vm will use the same vm and if any change needed we can do it with simple edit and then revert to the original state.
The other tests doesn't need the migration policy label, therefore restart after editing with resource editor will be faster then create windows vm with the policy label for just avoiding another restart which takes few more seconds.

File: tests/observability/metrics/conftest.py [hunk 13]
```diff
@@ -1058,9 +1059,38 @@ def windows_vm_for_test(namespace, unprivileged_client):
...
+                "spec": {
+                    "template": {"spec": {"nodeSelector": get_node_selector_dict(node_selector="non-existent-node")}}
+                }
+            }
+        }
+    ):
+        windows_vm_for_test.restart()
+        windows_vm_for_test.wait_for_specific_status(status=VirtualMachine.Status.ERROR_UNSCHEDULABLE)
+        yield windows_vm_for_test
+    windows_vm_for_test.restart()
```
Comment (rnetser) [inline tests/observability/metrics/conftest.py:1080]: doesn't the restart happen before ResourceEditor restores the spec?

File: tests/observability/metrics/conftest.py [hunk 14]
```diff
@@ -1058,9 +1059,38 @@ def windows_vm_for_test(namespace, unprivileged_client):
         vm_name="win-vm-for-test",
         storage_class=py_config["default_storage_class"],
     ) as vm:
+        running_vm(vm=vm)
         yield vm
 
 
+@pytest.fixture()
+def non_existent_node_windows_vm(windows_vm_for_test):
```
Comment (rnetser) [inline tests/observability/metrics/conftest.py:1067]: why do you need windows vm for this test (or any other negative test)? use a small image to save time

File: tests/observability/metrics/conftest.py [hunk 15]
```diff
@@ -1149,6 +1179,103 @@ def migration_succeeded_scope_class(vm_migration_metrics_vmim_scope_class):
...
+
+@pytest.fixture()
+def stopped_windows_vm(windows_vm_for_test):
+    windows_vm_for_test.stop()
+    yield windows_vm_for_test
+    running_vm(vm=windows_vm_for_test)
+
+
+@pytest.fixture(scope="module")
+def pvc_with_non_existent_pv(namespace, admin_client):
```
Comment (rnetser) [inline tests/observability/metrics/conftest.py:1195]: same applies here - one test (regardless of os) can be written for this use case

File: tests/observability/metrics/conftest.py (hunk 13 above)
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py:1080]: Yes, but I need another one after the ResourceEditor restores the sepc.

File: tests/observability/metrics/conftest.py (hunk 14 above)
Comment (rnetser) [reply tests/observability/metrics/conftest.py:1067]: iiuc the tested metrics here do not depend on the guest OS. right?

File: tests/observability/metrics/conftest.py (hunk 12 above)
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py:1080]: it Actually saves only one restart, Because from what you are suggesting I can remove the fixture `windows_vm_with_low_bandwidth_migration_policy` that doing one restart. 
If I will remove the restarts in  `non_existent_node_windows_vm` the test will fail and the tests after it also will fail, as I checked it multiple times.

**Comment by openshift-virtualization-qe-bot-3** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

File: tests/observability/metrics/conftest.py (hunk 12 above)
Comment (dshchedr) [reply tests/observability/metrics/conftest.py:1080]: > so is it worth to add this label to the windows vm for all the tests that it used only once just for saving one restart?

by adding label permanently you are saving time on a minimum 3 actions: `add label to vm -> restart vm -> remove label from vm`


but I can just repeat one of my previous comments, with the right tests order you can save much more time on unnecessary actions:


1. `test_vm_running_status_metrics_windows`    - VM can be already running from your previous tests, if not - create and run it

2. `test_vm_migrating_status_metrics_windows`     - if the label already added to the VM - no need to restart the VM. Just create migration policy, and migrate VM. 

3. `test_vm_non_running_status_metrics_windows`    - Since in the previous test you don't care if the migration completed - you can simply stop the VM

4. `test_vm_error_status_metrics_windows`       -  VM already stopped in previous test, so just edit it and start - it should be quick test because you don't wait when it is active anyway. During teardown - restart the VM and wait when it is running

**Comment by openshift-virtualization-qe-bot-3** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

**Comment by openshift-virtualization-qe-bot-3** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

**Comment by openshift-virtualization-qe-bot-3** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

File: tests/observability/metrics/conftest.py [hunk 16]
```diff
@@ -1148,6 +1168,102 @@ def migration_succeeded_scope_class(vm_migration_metrics_vmim_scope_class):
...
+        client=unprivileged_client,
+    ) as vm:
+        running_vm(vm=vm, wait_for_interfaces=False, check_ssh_connectivity=False)
+        yield vm
+
+
+@pytest.fixture(scope="class")
+def number_of_running_vmis(admin_client):
+    return len(list(VirtualMachineInstance.get(dyn_client=admin_client)))
+
```
[the remark this answers is not in the document]
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py:1265]: But this is the purpose of this fixture

**Comment by openshift-virtualization-qe-bot-3** [comment]
New container for quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166 published

File: tests/observability/metrics/conftest.py (hunk 10 above)
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py]: Without this restart at the end all the tests that use this windows vm will fail.

File: tests/observability/metrics/conftest.py (hunk 14 above)
Comment (OhadRevah) [reply tests/observability/metrics/conftest.py:1066]: @rnetser For this test class no.

**Comment by OhadRevah** [comment]
After talking with dev, metrics that not guest related should not be covered

**Comment by openshift-virtualization-qe-bot-2** [comment]
Successfully removed PR tag: quay.io/openshift-cnv/openshift-virtualization-tests:pr-1166.
